feat: context aware tools

ahmedrowaihi opened the pull request:
## 🎯 Changes

Supporting context in tool execution..

to make tools behaviour context aware, the only way is to reconstruct them everytime you want to invoke a new trigger..
this PR suggests supporting context.

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/ai/blob/main/CONTRIBUTING.md).
- [x] I have tested this code locally with `pnpm run test:pr`.
- [x] I have tested this code locally with `pnpm run test:lib`.

## 🚀 Release Impact

- [ ] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] This change is docs/CI/dev-only (no release).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Tools now receive context during execution through an options parameter, available in both client and server environments
  * Context can be passed via chat client options and is propagated to all tool executions

* **Tests**
  * Added and updated tests to verify context propagation to tool executions

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jherr commented on packages/typescript/ai/src/tools/tool-definition.ts:
```diff
@@ -177,27 +184,31 @@ export function toolDefinition<
+    client<TContext = unknown>(
       execute?: (
         args: z.infer<TInput>,
+        context?: TContext,
```
And once more `options`.

jherr reviewed (changes requested):
One small API change, otherwise I love it!

ahmedrowaihi commented:
@AlemTuzlak , do you think we better also pass current _serialisable_only_ within the request ?

ahmedrowaihi committed:
feat: tools context

feat: introduce ToolOptions and update tool execution context

fix: update context handling in ChatClient and API route
